### PR TITLE
Refactor/database and leaderboard

### PR DESCRIPTION
**Pull request by kangpl**

### CODE REVIEW

File: app/src/main/java/ch/epfl/sdp/database/firebase/api/ServerFirestoreDatabaseAPI.java [hunk 1]
```diff
@@ -79,13 +80,7 @@ public void sendItemBoxes(List<ItemBoxForFirebase> itemBoxForFirebaseList) {
 
     @Override
     public void sendPlayersHealth(List<PlayerForFirebase> playerForFirebaseList) {
-        sendPlayersProperty(playerForFirebaseList, "healthPoints", p -> p.getHealthPoints());
-    }
-
-
-    @Override
-    public void sendPlayersAoeRadius(List<PlayerForFirebase> playerForFirebaseList){
```
Comment (Zeavee) [inline app/src/main/java/ch/epfl/sdp/database/firebase/api/ServerFirestoreDatabaseAPI.java:87]: We do not need this method anymore?

File: app/src/main/java/ch/epfl/sdp/database/room/AppDatabase.java [hunk 2]
```diff
@@ -12,20 +12,26 @@
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 
-@Database(entities = {LeaderboardEntity.class}, version = 1)
+@Database(entities = {GeneralLeaderBoardEntity.class}, version = 2, exportSchema = false)
```
Comment (Zeavee) [inline app/src/main/java/ch/epfl/sdp/database/room/AppDatabase.java:15]: What does this change?

File: app/src/main/java/ch/epfl/sdp/game/StartGameController.java [hunk 3]
```diff
@@ -70,28 +75,36 @@ static Area initGameArea() {
...
-        randomEnemyGenerator.setEnemyCreationTime(1000);
-        randomEnemyGenerator.setMaxEnemies(10);
-        randomEnemyGenerator.setMinDistanceFromEnemies(100);
-        randomEnemyGenerator.setMinDistanceFromPlayers(100);
-        randomEnemyGenerator.generateEnemy(100);
-        Enemy enemy = randomEnemyGenerator.getEnemies().get(0);
-        enemyManager.updateEnemies(enemy);
-        //  -------------------------------------------
+    void generateEnemy(EnemyManager enemyManager) {
+        if(enemyManager.getEnemies().size() < MAX_ENEMY) {
```
Comment (Zeavee) [inline app/src/main/java/ch/epfl/sdp/game/StartGameController.java:93]: Isn't this this executed two times? Or was it moved here?

File: app/src/main/java/ch/epfl/sdp/database/room/AppDatabase.java (hunk 2 above)
Comment (kangpl) [reply app/src/main/java/ch/epfl/sdp/database/room/AppDatabase.java:15]: It seems I change the name of the table, then I need to upgrade the version.

File: app/src/main/java/ch/epfl/sdp/game/StartGameController.java (hunk 3 above)
Comment (kangpl) [reply app/src/main/java/ch/epfl/sdp/game/StartGameController.java:93]: This method just calls the method  generateEnemt in randomEnemyGenerator

File: app/src/main/java/ch/epfl/sdp/game/Server.java [hunk 4]
```diff
@@ -76,27 +70,31 @@ public void start() {
...
-        if (scoreTimeCounter <= 0) {
+        // Update the current game score in 10 seconds
+        if (counter % (10 * GameThread.FPS) == 0 ) {
             Log.d(TAG, "update: update the score of player");
             updateInGameScoreOfPlayer();
-            scoreTimeCounter = 10 * GameThread.FPS + 1;
         }
-        --scoreTimeCounter;
 
+        if (counter % (30 * GameThread.FPS) == 0) {
```
Comment (varathar53) [inline app/src/main/java/ch/epfl/sdp/game/Server.java:90]: perhabs you can add a javadoc here

File: app/src/main/java/ch/epfl/sdp/game/StartGameController.java [hunk 5]
```diff
@@ -26,23 +26,28 @@
 
 import static android.content.ContentValues.TAG;
 
-public interface StartGameController {
+public abstract class StartGameController {
+    private RandomEnemyGenerator randomEnemyGenerator;
+    private static final int MAX_ENEMY = 10;
```
Comment (varathar53) [inline app/src/main/java/ch/epfl/sdp/game/StartGameController.java:31]: perhabs you can write down that its in meter the distance.

File: app/src/main/java/ch/epfl/sdp/game/Server.java (hunk 4 above)
Comment (kangpl) [reply app/src/main/java/ch/epfl/sdp/game/Server.java:90]: For the code climate, it makes no sense why the code climate think they are duplicated, so we just ignore.

File: app/src/main/java/ch/epfl/sdp/game/StartGameController.java (hunk 5 above)
Comment (kangpl) [reply app/src/main/java/ch/epfl/sdp/game/StartGameController.java:31]: yeah

**Pull request merged**